lib/Core/BuildEngine.cpp

```diff
@@ -931,31 +942,97 @@ class BuildEngineImpl {
+            return false;
+          }
+        }
 
-    // If there was no work to do, but we still have running tasks, then
-    // we have found a cycle and are deadlocked.
-    if (!taskInfos.empty()) {
-      reportCycle(buildKey);
-      return false;
+        // We didn't do any work, and we have nothing more we can/need to do.
+        break;
+      }
     }
 
     return true;
   }
 
-  /// Report the cycle which has called the engine to be unable to make forward
+  /// Attempt to resolve a cycle which has called the engine to be unable to make forward
   /// progress.
   ///
   /// \param buildKey The key which was requested to build (the reported cycle
   /// with start with this node).
-  void reportCycle(const KeyType& buildKey) {
+  /// \returns True if the engine should try to proceed, false if the build the could not
+  /// be broken.
+  bool resolveCycle(const KeyType& buildKey) {
     // Take all available locks, to ensure we dump a consistent state.
     std::lock_guard<std::mutex> guard1(taskInfosMutex);
     std::lock_guard<std::mutex> guard2(finishedTaskInfosMutex);
 
+    std::vector<Rule*> cycleList = findCycle(buildKey);
+    assert(!cycleList.empty());
+
+    if (delegate.shouldResolveCycle(cycleList)) {
+      TracingInterval i(EngineQueueItemKind::BreakingCycle);
+
+      for (auto ruleIt = cycleList.rbegin(); ruleIt != cycleList.rend(); ruleIt++) {
+        auto& ruleInfo = getRuleInfoForKey((*ruleIt)->key);
+
+        // If this rule is scanning, try to force a rebuild to break the cycle
+        if (ruleInfo.isScanning()) {
+          if (trace) {
+            trace->cycleForceRuleNeedsToRun(&ruleInfo.rule);
+          }
+          for (auto it = ruleInfosToScan.begin(); it != ruleInfosToScan.end(); it++) {
+            if (it->ruleInfo->keyID == ruleInfo.keyID) {
+              ruleInfosToScan.erase(it);
+            }
+          }
+          finishScanRequest(ruleInfo, RuleInfo::StateKind::NeedsToRun);
+          return true;
+        }
+
+        // Check if this rule has a (potentially) valid previous result and if so
+        // try to provide it to the node requesting it to break the cycle.
+        if (ruleInfo.isInProgressWaiting() && ruleInfo.result.builtAt != 0) {
+          auto* taskInfo = ruleInfo.getPendingTaskInfo();
+
+          // find downstream node in the cycle that wants this input
+          auto& nextRuleInfo = getRuleInfoForKey((*std::next(ruleIt))->key);
+
+          // find the corresponding request on this task
+          for (auto it = taskInfo->requestedBy.begin(); it != taskInfo->requestedBy.end(); it++) {
+            if (it->taskInfo->forRuleInfo->keyID == nextRuleInfo.keyID) {
+              // supply the prior value to the node
+              if (trace) {
+                trace->cycleSupplyPriorValue(&ruleInfo.rule, it->taskInfo->task.get());
+              }
+              it->forcePriorValue = true;
+              finishedInputRequests.insert(finishedInputRequests.end(), *it);
```

Comment:
I think it is possible that this change exposes the following very subtle race:
1. I *think* it is true that we don't protect accesses to `ruleInfo->result`, because no task should ever try and read them before they have been written.
2. However, with this change, the processing of the finished input request will in fact try to read the result.
3. If the task meanwhile gets spun up and finishes, it could be writing that.

I think we might end up needing a separate queue to hold the "forced input requests" (since either copying the result or adding a lock to the hot path feels annoying for something which will rarely trigger).

Can you confirm my reasoning here? I may have missed something.

```

Reply:
Indeed, this thought crossed my mind.  I convinced myself it was okay, but perhaps being explicit about it would be safer, or at least a comment explaining the logic. Basically, if we have reached this point, the engine has stalled all other work.  We are explicitly unblocking only a single item with this prior value at this point, thus it should be the first task executed in the main task loop above.